### ISSUE TITLE
[selectors-4] Introduce :popover-target-open pseudo class

### ISSUE THREAD

**Issue by ggedde** [w3c/csswg-drafts#9547]
Looks like :popover-open is landing, but it only applies to the popover element itself.
We also need a way to style the Trigger element for the popover element when the popover element is open.

consider the following:
```
<button popovertarget="my-popover">Open Popover</button>
...
<div popover id="my-popover">Greetings, one and all!</div>
```

I would like a way to style the button when the popover is open.
This would need to apply to all elements the the same 'popovertarget'
Ex.

```
<button popovertarget="my-popover">Open Popover</button>
...
<div popover id="my-popover">Greetings, one and all!</div>
...
<button popovertarget="my-popover">Open Popover From another button</button>
```

It would be ideal to have something like:
```
button:popover-target-open {
   ...
}
```

Thanks

**Comment by YummyBacon5** [comment]
`popovertarget` is being planned to be replaced with `invoketarget` by the Open UI CG. ([explainer](https://open-ui.org/components/invokers.explainer/))

Meaning that the name may be different. And that other `invokeaction`s could be included. As well as `interesttarget`

**Comment by ggedde** [comment]
Looks like the explainer for invoketarget might use `aria-expanded=true` which is good. Doesn't matter to me as long as we have a way to handle this.

**Comment by YummyBacon5** [comment]
It would in the accessibility API. But this wouldn't be an attribute which could be selected with CSS

**Comment by nt1m** [comment]
cc @keithamus if you have ideas of what can be done here.

I would think :has(:popover-open) + a certain DOM structure makes this possible. Having something more generic for invokers would require a precise definition.

**Comment by ggedde** [comment]
> It would in the accessibility API. But this wouldn't be an attribute which could be selected with CSS. So a new pseudo-class selector for this would still be needed

I am not that familiar with the accessibility API, but if `aria-pressed` is updated then we can access that with CSS. Do you have any links or information showing this in the accessibility API?
Thanks

**Comment by ggedde** [comment]
> cc @keithamus if you have ideas of what can be done here.
> 
> @nt1m I would think :has(:popover-open) + a certain DOM structure makes this possible. Having something more generic for invokers would require a precise definition.

We shouldn't have to rely on adding the elements next to each other or in a specific container, etc. The power of the popover api allows us to have the elements separate from each other. At which point we would need a way to know if the invokers target has been changed. As far as a precise definition as long as they contain `popovertarget` then we can determine if the target is open then add a pseudo class to the invoker. Essentially, whenever a target has changed the browser should fine ALL invokers and add a pseudo class to them as well.
Thoughts?

**Comment by nt1m** [comment]
With the invokers proposal, I wonder if we should have a :invoketarget() function that takes a selector to match against.

`button:invoketarget(:popover-open)` would match a button whose invoketarget matches `:popover-open`.

Implementing invalidation for something like this would probably involved though (imagine all the combinations like `:invoketarget(:has(...))`!)

**Comment by YummyBacon5** [comment]
> I am not that familiar with the accessibility API, but if aria-pressed is updated then we can access that with CSS. Do you have any links or information showing this in the accessibility API?

Every HTML element is internally mapped to ARIA role and state(s). Where these properties are then communicated to assistive technologies through an [accessibility API](https://w3c.github.io/aria/#dfn-accessibility-api).

But basically, the actual ARIA attributes are not going to be added to the elements, and are just represented internally. So a CSS selector can't select them.

---

> `button:invoketarget(:popover-open)` would match a button whose invoketarget matches `:popover-open`.

This was also suggested by @lukewarlow in https://github.com/openui/open-ui/issues/898#issuecomment-1783952332

**Comment by lukewarlow** [comment]
Hmm @nt1m your function selector is probably a much better idea than my pseudo element one (especially because pseudo elements don't currently work in :has()).

I do feel like some way to select in css is needed though. Because it's common for playpause buttons for example to change their appearance based on state and that would currently rely on a specific Dom structure and has().

But as you say it would be quite picky with regards to invalidation.

Could we potentially flip the idea?

`#popover:popover-open:invoker()` or `#popover:popover-open::invoker`

Would this simplify things?

**Comment by keithamus** [comment]
Perhaps a more generic selector which applies when an element that is tied to another open element is expanded - effectively providing a styling hook for the implicit `aria-expanded` state. Let's say `:expanded`. This would apply to the button that expanded an element, and would be applied to buttons with a `popovertarget` that points to an open popover, or a `invoketarget` that points to an open popover, open dialog, or open details element. It could also apply to Summary elements that have open details (so effectively `summary:expanded` and `details[open] > summary` would be the same thing), and would be helpful for the new `<selectlist>` and `<combobox>` elements.

**Comment by YummyBacon5** [comment]
> Perhaps a more generic selector which applies when an element that is tied to another open element is expanded - effectively providing a styling hook for the implicit `aria-expanded` state

This would be confusing for the video pause/play icon use case. Since saying that a button for a video is "expanded" just because the video is playing is confusing.

Also I don't think that this would work with custom invoker actions - where there's no true/false state.

**Comment by lukewarlow** [comment]
Yeah that would only be doable for the popover, dialog, details scenario
. Not all invokers, which perhaps is fine? Maybe it's something worth thinking on separately from a more general invokers specific one?

**Comment by keithamus** [comment]
> > Perhaps a more generic selector which applies when an element that is tied to another open element is expanded - effectively providing a styling hook for the implicit `aria-expanded` state
> 
> This would be confusing for the video pause/play icon use case. Since saying that a button for a video is "expanded" just because the video is playing is confusing.
> 
> Also I don't think that this would work with custom invoker actions - where there's no true/false state.

Both of these are correct and intentional.

- Video pause/play would not be `aria-expanded` but instead be `aria-pressed`. We could have another pseudo class for this case but I'm not fully convinced there's a use case that differentiates from other states like `:active`.
- We _cannot_ determine if custom invoker actions influence the`aria-pressed` or `aria-expanded` state, and so in these cases the very likely outcome is that browsers won't expose any implicit aria information as a consequence of invoking a custom invoker action. Consequently your application scripting would need to deal with this, which likely means `setAttribute('aria-pressed')` or `setAtrribute('aria-expanded')`. In those cases the attributes become easy to select for, meaning that a selector for `button:is(:expanded, [aria-expaned="true"])` would cover all expanded cases.

**Comment by o-t-w** [comment]
Why not reuse the pre-existing `:open` pseudo-class? https://drafts.csswg.org/selectors-4/#open-state

```css
button:open
```
